### ISSUE TITLE
Viewer handler keeps reference to Navigator

### ISSUE THREAD

**Issue by zero41120** [openseadragon/openseadragon#1721]
https://github.com/openseadragon/openseadragon/blob/3bbfbcbc0e136405e3f8ed95316d36255dba835f/src/navigator.js#L399

https://github.com/openseadragon/openseadragon/blob/3bbfbcbc0e136405e3f8ed95316d36255dba835f/src/navigator.js#L215

My project has 2 active navigators. One is regular navigator that shows everything. 
The other is navigator for the current image that user is viewing.

My project has custom logic that detect when user is viewing an image at a particular zoom level and creates a navigator for it by 
`viewer.slidesNavigator = new OpenSeadragon.Navigator(options);`

By taking memory snapshots, we found that handlers in the main viewer keeps a reference navigators that are no longer in use. 
![image](https://user-images.githubusercontent.com/10471270/66873378-1f70e480-ef5d-11e9-87a9-3b2697fcd4d9.png)
![image](https://user-images.githubusercontent.com/10471270/66873726-0583d180-ef5e-11e9-8fea-3e07bc72cd13.png)

Thus the navigator will never be garbage collected

**Comment by iangilman** [comment]
Good catch! Do you have a sense for where this is happening? Would you be interested in making a patch for it?

**Comment by zero41120** [comment]
Our current workaround is to unset the handlers to empty array before we add new navigator instance.
We know where we create new navigator so we can unset it safely.

Since navigator extends the viewer, possible solution is to update .destroy() method for navigator to remove the handlers associated with the navigator.

**Comment by iangilman** [comment]
Yes, that sounds like a good solution!

**Comment by zero41120** [comment]
@iangilman 
I want to informed you that I'm probably not the best person to fix to this issue.
I lack the knowledge of the OSD infrastructure and might accidentally break the functionality.

**Comment by zero41120** [comment]
@iangilman 
I was trying to recreate the issue without my project's context.
So I wrote the following file to simulate a intentional workflow (Using .destroy on navigator and viewer)
https://github.com/zero41120/openseadragon/commit/1257836a0e06948af516002b19d0f2b5ee868b6d

I do not see the rotate events accumulating, instead, mouse tracker event now holds the reference
![image](https://user-images.githubusercontent.com/10471270/68934662-f5713480-074b-11ea-9d83-dc0d0afe7bcd.png)

**Comment by iangilman** [comment]
Hmm... it looks like our `Viewer.destroy` needs to clean up more than it does, regardless of navigators. I've cut down your example (very helpful, BTW) to just create and then destroy a viewer:

```
        const viewer = new OpenSeadragon.Viewer({
            id: "mainOsd",
            tileSources,
            collectionMode: true,
            collectionRows: 2,
            showNavigator: false,
        });
        
        viewer.addHandler('open', function() {
          viewer.destroy();
        });
```

...and there's still a viewer in the Chrome memory snapshot (cool tool, BTW... I hadn't played with it before). It's referenced in quite a number of places: 

![image](https://user-images.githubusercontent.com/338989/68964309-de533680-078d-11ea-844b-c1df286178f6.png)

So, if we really want to get to the bottom of this, my recommendation would be to start on cleaning up the `Viewer.destroy` even before getting into the navigator.

As for your scenario, I think you can probably get away with not having to destroy and recreate the viewers. At least in this cut down version you provided, you can achieve the same effect like so:

```
        function setupSingleNavigator(tileSourceIndex) {
            hiddenViewer.open(tileSources[tileSourceIndex]);
        }
```

**Comment by JoFrMueller** [comment]
Thanks @iangilman for pointing this out. I just did land on this issue after realizing the leaky behaviour in Chrome as well.

As we're constantly navigating from/to SPA routes that load viewers (single/multiple) with different image material, we're unable to just call 'open' on the viewer object but would have to reset all paramaters accordingly.

I'd be really happy to help with a complete destruction of all references as we would have to hack around some kind of 'viewer pool' otherways. I started diving into the source code of OSD because of #1807 but feel a little bit lost when it comes to the destruction topic.

Could you help me out and share further information regarding this topic? Or are there any news on this already?

**Comment by JoFrMueller** [comment]
I made a first attempt to fix the leaks within #1808.

**Comment by iangilman** [comment]
@JoFrMueller Wonderful... thank you for jumping on this!
